pybop/optimisers/_cuckoo.py

```diff
@@ -0,0 +1,197 @@
+    of some cuckoo species. This algorithm was introduced by Yang and Deb in 2009.
+
+    The algorithm uses a population of host nests (solutions), where each cuckoo
+    (new solution) tries to replace a worse nest in the population. The quality
+    or fitness of the nests is determined by the cost function. A fraction
+    of the worst nests is abandoned at each generation, and new ones are built
+    randomly.
+
+    The pseudo-code for the Cuckoo Search is as follows:
+
+    1. Initialise population of n host nests
+    2. While (t < max_generations):
+        a. Get a cuckoo randomly by Lévy flights
+        b. Evaluate its quality/fitness F
+        c. Choose a nest among n (say, j) randomly
+        d. If (F > fitness of j):
+            i. Replace j with the new solution
+        e. Abandon a fraction (pa) of the worst nests and build new ones
+        f. Keep the best solutions/nests
+        g. Rank the solutions and find the current best
+    3. End While
+
+    This implementation also uses a decreasing step size for the Lévy flights, calculated
+    as sigma = sigma0 / sqrt(iterations), where sigma0 is the initial step size and
+    iterations is the current iteration number.
+
+    Parameters:
+    - pa: Probability of discovering alien eggs/solutions (abandoning rate)
+
+    References:
+    - X. -S. Yang and Suash Deb, "Cuckoo Search via Lévy flights,"
+      2009 World Congress on Nature & Biologically Inspired Computing (NaBIC),
+      Coimbatore, India, 2009, pp. 210-214, https://doi.org/10.1109/NABIC.2009.5393690.
+
+    - S. Walton, O. Hassan, K. Morgan, M.R. Brown,
+      Modified cuckoo search: A new gradient free optimisation algorithm,
+      Chaos, Solitons & Fractals, Volume 44, Issue 9, 2011,
+      Pages 710-718, ISSN 0960-0779,
+      https://doi.org/10.1016/j.chaos.2011.06.004.
+    """
+
+    def __init__(self, x0, sigma0=0.01, boundaries=None, pa=0.25):
+        super().__init__(x0, sigma0, boundaries=boundaries)
+
+        # Problem dimensionality
+        self._dim = len(x0)
+
+        # Population size and abandon rate
+        self._n = self._population_size
+        self._pa = pa
+        self.step_size = self._sigma0
+        self.beta = 1.5
+
+        # Set states
+        self._running = False
+        self._ready_for_tell = False
+
+        # Initialise nests
+        if self._boundaries:
+            self._nests = np.random.uniform(
+                low=self._boundaries.lower(),
+                high=self._boundaries.upper(),
+                size=(self._n, self._dim),
+            )
+        else:
+            self._nests = np.random.normal(
+                self._x0, self._sigma0, size=(self._n, self._dim)
+            )
+
+        self._fitness = np.full(self._n, np.inf)
+
+        # Initialise best solutions
+        self._x_best = np.copy(x0)
+        self._f_best = np.inf
+
+        # Set iteration count
+        self._iterations = 1
```

Comment:
I'm not sure if this should count as the first iteration, with an infinite cost?
I see that `self._iterations` is used in line 97 below, but maybe set it equal to zero here and add 1 to line 97 for consistency.